### ISSUE TITLE
[expo.google.logInAsync] Wrapper code doesn't take standalone clientIds when it should

### ISSUE THREAD

**Issue by gerbus** [expo/expo#3253]
See [this thread from the forum](https://forums.expo.io/t/sdk-32-google-login-android-error-redirect-uri-mismatch/17976/15?u=gerbus)

### Steps to Reproduce

1. Call `logInAsync`, and provide with both non-standalone and standalone clientIds in params. 

Note importantly that non-standalone OAuth credentials have `host.exp.exponent` as the package name, and standalone credentials have something different (like `com.wallstreetjournal`). If these don't match the package name (specified in `app.json` for standalone) related to the requester (who is either `expo` or `standalone`, as indicated in `Expo.Constants.appOwnership`), I believe google correctly gives a `request_uri_mismatch` error.

```
  try {
    resp = await Expo.Google.logInAsync({
      behavior: "web",
      scopes: [
        'https://www.googleapis.com/auth/userinfo.email',
        'https://www.googleapis.com/auth/userinfo.profile',
        'https://www.googleapis.com/auth/calendar',
        'https://www.googleapis.com/auth/calendar.events',
      ],
      androidClientId: GOOGLE_CLIENTID_ANDROID_EXPO,
      iosClientId: GOOGLE_CLIENTID_IOS_EXPO,
      androidStandaloneAppClientId: GOOGLE_CLIENTID_ANDROID_STANDALONE,
      iosStandaloneAppClientId: GOOGLE_CLIENTID_IOS_STANDALONE
    });
}
```
2. Do a standalone build
3. Perform a login in the standalone build.

### Expected Behavior
The app opens the system web browser and the google login screen is shown

### Actual Behavior
The app opens the system web browser and gives this error message:

![screenshot_chrome_20190118-085818](https://user-images.githubusercontent.com/2348366/51401527-da1fa380-1aff-11e9-919f-528deee126a4.png)

I checked the `clientId` in the error, and it isn't my standalone clientId (which it should be), it's the expo clientId. Since I did a standalone build, the requestor's package name for the google login comes from my `app.json` (`com.wallstreetjournal`). Since the clientId provided is tied to the package name `host.exp.exponent`, there's a mismatch.

I think the [problem code is on lines 62-63 here](https://github.com/expo/expo/blob/cfe899cc8a59eb892a193a94c54890b3c8ee160c/packages/expo/src/Google/Google.ts#L62). It always grabs the non-standalone clientIds. Maybe someone assumed (as I did, at first) that since the `system` behavior is deprecated, the standalone clientIds aren't needed. They are!.

**Comment by EvanBacon** [comment]
You are correct, the docs need to be updated. For information on `system` authentication, please refer to this article: https://blog.expo.io/react-native-google-sign-in-with-expo-d1707579a7ce

**Comment by ansonhoyt** [comment]
@EvanBacon is this still in transition? The [docs](https://docs.expo.io/versions/latest/sdk/google/), [code](https://github.com/expo/expo/blob/master/packages/expo/src/Google/Google.ts), and [blog post](https://blog.expo.io/react-native-google-sign-in-with-expo-d1707579a7ce) still don't seem to match up with each other.

Like @gerbus, I'm trying to authenticate with a Google API in Expo SDK 32 for a standalone app. I'd like to use `behavior: system`, making it transparent to the user. It seems that Expo's Google module is still forcing `behavior: web` even when we force in the standalone clientId.

I'm a bit confused, but if I follow https://github.com/expo/expo/pull/2866, I should be writing a wrapper around Expo [AppAuth](https://github.com/expo/expo/tree/master/packages/expo-app-auth) instead of using [Google](https://docs.expo.io/versions/latest/sdk/google/). Is that right?

**Comment by EvanBacon** [comment]
TL;DR: follow my [blog post](https://blog.expo.io/react-native-google-sign-in-with-expo-d1707579a7ce) much easier to release and edit these 🙃 
@ansonhoyt really sorry for the run around. The situation is something like this:
Google native auth never worked in the client, the API was somewhat confusing because Google web auth is very advanced and looks like the native auth.

Then Google auth changed to work better with Firebase, which removed any chance to sandbox the functionality. 

To combat this I created 2 libraries which essentially did the same thing `Expo.Google` was assuming.

1. expo-app-auth: a wrapper on the native OAuth library `AppAuth` which can be used in the Expo Client. This is a very good solution and is identical to the functionality users of `Expo.Google` were getting. (aside from a bug on Android which requires you to run UNVERSIONED as the SDK value (this should be fixed soon)). `Expo.Google` is a JS wrapper around this library. To use it you should only pass your `clientId` to the sign-in method.
2. expo-google-sign-in: a wrapper on the native `GoogleSignIn` library. This module can only be used by configuring your `app.json` in a special way, then building and running on in a production app. Or by ejecting to the unmanaged workflow and adding the Google REVERSE_CLIENT_ID to your URL schemes on iOS. (I personally detach on a branch and test the functionality this way)

I'll get around to updating the docs soon, unfortunately I'm just really stretched thin. 😕

**Comment by EvanBacon** [comment]
The docs have now been updated: 
* In Client: https://docs.expo.io/versions/v32.0.0/sdk/google/
* In Standalone: https://docs.expo.io/versions/v32.0.0/sdk/google-sign-in/

**Comment by ansonhoyt** [comment]
I'll give it a whirl. Many thanks for closing this loop amidst all the other things you're working on!

**Comment by vietnogi** [comment]
@EvanBacon for `Expo.Google`, is the `clientId` the web client ID for oauthclient?  Thats what I am using and I get Error: ```invalid_request - Custom scheme URIs are not allowed for `WEB` client type.```

**Comment by victor9000** [comment]
@vietnogi I ran into this problem when following [these instructions](https://docs.expo.io/versions/v32.0.0/sdk/google/#server-side-apis) and running my app on a physical device.  Google (unlike Facebook) is very particular about the client matching the specified environment, so you'll need to create an OAuth client that matches your device type (Android or iOS).

**Comment by edoantonioco** [comment]
can [this](https://docs.expo.io/versions/v32.0.0/sdk/google/#server-side-apis) instructions please be correct?

Like try them and if they dont work then fix the instructions to a way that they works.

**Comment by urosran** [comment]
**Tried setting up Google auth in Client such as:** 

import React from 'react';
import { ScrollView, StyleSheet, Text, Button } from 'react-native';
import {Google} from 'expo';

export default class LoginScreen extends React.Component {
  static navigationOptions = {
    title: 'LoginScreen',
  };

  onPressHandler = async () => {
    const clientId = '649846420860-nd92louc1phcmh8h68vofonfk8mo7al1.apps.googleusercontent.com';
    const { type, accessToken, user } = await Google.logInAsync({ clientId });

    if (type === 'success') {
      console.log(user);
    }
  };

  render() {
    return (
      <ScrollView style={styles.container}>
        <Button
        title="Google"
        onPress={() => this.onPressHandler()}/>
      </ScrollView>
    );
  }
}

**When running this in android emulator on win10 and expo app in IOS the same error appears:** 
![image](https://user-images.githubusercontent.com/25808647/55295404-a2c05380-53da-11e9-837d-ee69e013f25b.png)

**Comment by robertobrogi** [comment]
@urosran i have the same problem..... have you fix?? there is more confusion in this new approach....
there is any fix for the Google.logInAsync ???

**Comment by robertobrogi** [comment]
i find a fix!

behavior: 'web',
                scopes: ['profile', 'email'],
                androidClientId: Expo.Constants.appOwnership === 'standalone' ? PROPERTIES.ANDROID_STAND_ALONE_APP_CLIENT_ID : PROPERTIES.ANDROID_CLIENT_ID,
                iosClientId: Expo.Constants.appOwnership === 'standalone' ? PROPERTIES.IOS_STAND_ALONE_APP_CLIENT_ID : PROPERTIES.IOS_CLIENT_ID,
                androidStandaloneAppClientId: PROPERTIES.ANDROID_STAND_ALONE_APP_CLIENT_ID,
                iosStandaloneAppClientId: PROPERTIES.IOS_STAND_ALONE_APP_CLIENT_ID,
                webClientId: '274334129691-k9shjbjjttr3tfj39opodgr2llusucn0.apps.googleusercontent.com'

thanks all

**Comment by kamalpandey** [comment]
@robertobrogi  I have changed as you mentioned but it's not working

**Comment by kamalpandey** [comment]
Finally Solved using google-sign-in for standalone

**Comment by tan116** [comment]
Hi I have a problem with the new google sign in on sdk 32. I first got the redirect uri mismatch. In the process of upgrading, I deleted an androidClientId that I was using previously. And I started seeing "deleted_client" upon google sign in.

But then I created a new one and coded it in. Still didn't work. And now, even after changing all my code to the new version as suggested  [here](https://blog.expo.io/react-native-google-sign-in-with-expo-d1707579a7ce), the google login still throws up the same error - deleted_client. The client_id in the error message is the one that got deleted and I have tried everything but I am not able to fix this.

Does anyone have an idea on what I could do?

**Comment by Milankorangi** [comment]
use host.exp.exponent as bundler identifier and try it

**Comment by tan116** [comment]
@EvanBacon Can you please clarify two things for me -

1. From sdk 32 onward, can we use Google.loginAsync() in a way that it works on both expo client and standalone apps simultaneously like before?
2. If not, does that mean using GoogleSignIn is necessary for standalone apps?

Doing an update for the prod version of my app and these two answers would really help me plan my work. Thanks!

**Comment by Milankorangi** [comment]
@tan116  
1) yeah above 32 it only works for android expo client
2) yeah its certain to use GoogleSignIn for standalone

**Comment by jeffreybrowning** [comment]
@EvanBacon it would be great to get a write-up on how you test google sign in by ejecting. I am attempting to have `webClientId` set for use on a python backend server and am resorting to testing by ejecting.

**Comment by cruzach** [comment]
I just tested `Google.logInAsync` on SDK 35 on both standalone and Expo client, following the steps in the [docs](https://docs.expo.io/versions/latest/sdk/google/#using-it-inside-of-the-expo-app) exactly. The configuration required for this one has more than a few steps, so if you're having trouble I'd assume it's due to that. There was a fix for this library a while back that I believe must have solved this, but if you are still seeing this issue on an SDK we support, please open up a new issue, thanks!
